src/storage/raw/raw_mvcc.rs

```diff
@@ -0,0 +1,199 @@
+        &self,
+        cf: Option<CfName>,
+        opts: Option<ReadOptions>,
+        key: &Key,
+    ) -> Result<Option<Value>> {
+        let mut iter_opt = IterOptions::default();
+        iter_opt.set_fill_cache(opts.map_or(true, |v| v.fill_cache()));
+        iter_opt.use_prefix_seek();
+        iter_opt.set_prefix_same_as_start(true);
+        let end_key = key.clone().append_ts(TimeStamp::zero());
+        iter_opt.set_upper_bound(end_key.as_encoded(), DATA_KEY_PREFIX_LEN);
+        let mut iter = match cf {
+            Some(cf_name) => self.iter_cf(cf_name, iter_opt)?,
+            None => self.iter(iter_opt)?,
+        };
+        if iter.seek(key)? {
+            Ok(Some(iter.value().to_owned()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<S: Snapshot> Snapshot for RawMvccSnapshot<S> {
+    type Iter = RawMvccIterator<S::Iter>;
+    type Ext<'a>
+    where
+        S: 'a,
+    = S::Ext<'a>;
+
+    fn get(&self, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(None, None, key)
+    }
+
+    fn get_cf(&self, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(Some(cf), None, key)
+    }
+
+    fn get_cf_opt(&self, opts: ReadOptions, cf: CfName, key: &Key) -> Result<Option<Value>> {
+        self.seek_first_key_value_cf(Some(cf), Some(opts), key)
+    }
+
+    fn iter(&self, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter(iter_opt)?))
+    }
+
+    fn iter_cf(&self, cf: CfName, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter_cf(cf, iter_opt)?))
+    }
+
+    #[inline]
+    fn lower_bound(&self) -> Option<&[u8]> {
+        self.snap.lower_bound()
+    }
+
+    #[inline]
+    fn upper_bound(&self) -> Option<&[u8]> {
+        self.snap.upper_bound()
+    }
+
+    fn ext(&self) -> S::Ext<'_> {
+        self.snap.ext()
+    }
+}
+
+pub struct RawMvccIterator<I: Iterator> {
+    inner: I,
+    cur_key: Option<Vec<u8>>,
+    cur_value: Option<Vec<u8>>,
+    is_valid: Option<bool>,
+}
+
+fn is_user_key_eq(left: &[u8], right: &[u8]) -> bool {
+    let len = left.len();
+    if len != right.len() {
+        return false;
+    }
+    Key::is_user_key_eq(left, &right[..len - U64_SIZE])
+}
+
+impl<I: Iterator> RawMvccIterator<I> {
+    fn new(inner: I) -> Self {
+        RawMvccIterator {
+            inner,
+            cur_key: None,
+            cur_value: None,
+            is_valid: None,
+        }
+    }
+
+    fn update_cur_kv(&mut self, key: Vec<u8>, val: Vec<u8>) {
+        self.cur_key = Some(key);
+        self.cur_value = Some(val);
+        self.is_valid = Some(true);
+    }
+
+    fn clear_cur_kv(&mut self) {
+        self.cur_key = None;
+        self.cur_value = None;
+        self.is_valid = None;
+    }
+
+    fn move_to_prev_max_ts(&mut self, res: Result<bool>) -> Result<bool> {
+        if *res.as_ref().unwrap() && self.inner.valid()? {
```

Comment:
Seems res is always unwrap, why not just take a bool instead?

Reply:
good idea. This let me rethink of the usage of unwrap and res here. I think res is not suitable to be an paramenter of this function. I remove it from this function.